controllers/storagecluster/external_ocs.go

```diff
@@ -0,0 +1,221 @@
+	}
+
+	instance.Status.ExternalStorage.ConsumerID = response.StorageConsumerUUID
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return reconcile.Result{}, nil
+}
+
+// offboardConsumer makes an API call to the external storage provider cluster for offboarding
+func offboardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	_, err := r.externalClusterClient.OffboardConsumer(instance.Status.ExternalStorage.ConsumerID)
+	if err != nil {
+		s, ok := status.FromError(err)
+		if !ok {
+			// This is not gRPC error. The operation must have failed before gRPC
+			// method was called, otherwise we would get gRPC error.
+			return reconcile.Result{}, err
+		}
+		r.logGrpcErrorAndReportEvent(instance, "OffboardConsumer", err, s.Code())
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// updateConsumerCapacity makes an API call to the external storage provider cluster to update the capacity
+func updateConsumerCapacity(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	response, err := r.externalClusterClient.UpdateCapacity(
+		instance.Status.ExternalStorage.ConsumerID,
+		instance.Spec.ExternalStorage.RequestedCapacity.String())
+	if err != nil {
+		s, ok := status.FromError(err)
+		if !ok {
+			// This is not gRPC error. The operation must have failed before gRPC
+			// method was called, otherwise we would get gRPC error.
+			return reconcile.Result{}, err
+		}
+		r.logGrpcErrorAndReportEvent(instance, "UpdateCapacity", err, s.Code())
+		return reconcile.Result{}, err
+	}
+
+	if response.GrantedCapacity == "" {
+		err = fmt.Errorf("External-OCS:UpdateCapacity:response is empty")
+		r.Log.Error(err, "empty response")
+		return reconcile.Result{}, err
+	}
+
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return reconcile.Result{}, nil
+}
+
+// getExternalConfigFromProvider makes an API call to the external storage provider cluster for json blob
+func getExternalConfigFromProvider(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) ([]ExternalResource, reconcile.Result, error) {
+
+	response, err := r.externalClusterClient.GetStorageConfig(instance.Status.ExternalStorage.ConsumerID)
+	if err != nil {
+		s, ok := status.FromError(err)
+		if !ok {
+			// This is not gRPC error. The operation must have failed before gRPC
+			// method was called, otherwise we would get gRPC error.
+			return nil, reconcile.Result{}, err
+		}
+		r.logGrpcErrorAndReportEvent(instance, "GetStorageConfig", err, s.Code())
+
+		// storage consumer is not ready yet, requeue after some time
+		if s.Code() == codes.Unavailable {
+			return nil, reconcile.Result{RequeueAfter: time.Second * 5}, nil
+		}
+
+		return nil, reconcile.Result{}, err
+	}
+
+	var externalResources []ExternalResource
+	err = json.Unmarshal(response.Data, &externalResources)
```

Comment:
Why do we need to unmarshal the data? I think we agreed that we are going to return the response a list of external resource objects, not as a blob of data.

Reply:
So we are using ExternalResource on the server-side for all operations and converting it into bytes just before sending the response to the client, that's why we need to unmarshal

Reply:
But that is my question, why are we doing it? what was the reason to convert into bytes in the first place where we have a structured response